### ISSUE TITLE
[Feature] Raise exception on "Execution context was destroyed"

### ISSUE THREAD

**Issue by Fjan** [microsoft/playwright#26740]
I'm trying to find places in my tests that emit "Execution context was destroyed, most likely because of a navigation".
According to Google Bard I should be able to use ```context = browser.new_context(throwOnContextDestroyed=True)```

However, I cannot find any reference in the documentation to that feature, so I assume it's a hallucination. Since Bard confidently predicts that this exists I can't be the only one pining for this. (Bard even provided me with a link to the documentation that also didn't exist 🙃)

**Comment by mxschmitt** [comment]
yup its referring to an option and entry in the documentation which does not exist. We saw related issues in the past which turned out to be Playwright bugs, so make sure you are using the latest version.

What problem are you encountering? Could you provide us a bit more details of what the issue is?

If you think its a Playwright bug, we need a reproducible in order to look into it, thanks!

**Comment by Fjan** [comment]
Thanks. I'm converting a large test suite from selenium to playwright, and to find where the warnings are generated I'm adding lots of debugging statements. When I find the offending line I can often fix the problem by waiting for something or adding a sleep (yuck). So, I'm pretty sure it's a problem with my tests rather than with Playwright but it's a laborious task to hunt for the right place. A way to flunk the test would be great to quickly find the offending lines.

**Comment by mxschmitt** [comment]
> waiting for something or adding a sleep (yuck).

Generally speaking there should be never a need for a sleep except if you want to ensure a certain element is not there after X seconds. See here for our [best-practises](https://playwright.dev/docs/best-practices), especially to use our [assertions](https://playwright.dev/docs/test-assertions) since they retry over and over again until the condition is met or the timeout is reached.

---

Is just the error getting displayed without a line where the error is coming from? This should not be the case usually. Which language are you using, your code above looks like Python?

**Comment by Fjan** [comment]
That error / warning gets emitted by playwright, the test runner I use (Capybara) is in a different process, so knowing the line number in Playwright probably wouldn't help me track down the test that's causing it. Having an option to throw an exception would break the test so that would reveal the line number.

I'm indeed aware sleeping is a very bad idea in a test suite. There's a few places I didn't manage to fix this warning using assertions when using playwright in combination with the webkit engine, with chromium I managed to get everything working, but that's something for another issue. It could be due to the ruby gems I'm using being relatively new.

**Comment by mxschmitt** [comment]
Which language binding is it tho? Seems something third-party and not supported by us, so I unfortunately need to close it then.

**Comment by Fjan** [comment]
The warning "Execution context was destroyed, most likely because of a navigation" is emitted by Playwright, that's not related to language bindings. This issue is a feature request to be able to pass a startup parameter to Playwright to raise an exception instead of a warning. 

Perhaps from my initial description it isn't clear this is a feature request? Apologies in that case.

**Comment by mxschmitt** [comment]
Thank you for clarifying. Yeah the feature request is unfortunately outside of the scope of Playwright (since it sounds more like a bug rather than a feature to me) and we officially don't support third-party language bindings. I recommend filing a bug report instead (follow our bug template) which uses an official language binding with a minimal reproduction attached, which surfaces the issue.

Thanks for our understanding!

**Comment by mxschmitt** [comment]
The feature you want is to hide an exception, which gets surfaced from Playwright internally. The exception should not get surfaced in the first place, so its a bug on our end. To fix this bug, we need to be able to reproduce it, find out why this is happening etc. In order to reproduce it, we'd appreciate if you could find out what exact lines are causing this for you and then, since the Playwright Ruby version is very similar to Python, create a small Python replica of the test, which should surface the same error. (any other language binding like Node.js works too)

I hope this makes it more clear, if you have more questions about the process, let us know, thanks!

**Comment by mitar** [comment]
While debugging a similar I found this issue. I think OP @Fjan was misunderstood. He was not asking for `browser.new_context(throwOnContextDestroyed=False)` to hide the exception, he was asking for `browser.new_context(throwOnContextDestroyed=True)` to throw an exception. I think the reasoning was that just "Execution context was destroyed, most likely because of a navigation" message lacks a stack trace so he hoped that if instead of a message there would be a proper exception, he would be able to debug the issue better.

So this is why this is a feature request and not bug report.

**Comment by Fjan** [comment]
Hi @mitar, yeah I could have formulated that better. Anyway, I solved my own issue by creating a little monkey patch; if you are using this with Rails & Capybara you can drop in the patch below to give you a stack trace when a warning occurs. 

```ruby
module Capybara
  module Playwright
    class Browser

      BC = ActiveSupport::BacktraceCleaner.new
      BC.add_filter   { |line| line.delete_prefix("#{Rails.root}/") }
      BC.add_silencer { |line| line =~ /\/lib\// }

      private def assert_page_alive(retry_count: 5, &block)
        if !@playwright_page || @playwright_page.closed?
          raise NoSuchWindowError
        end

        if retry_count <= 0
          return block.call
        end

        begin
          return block.call
        rescue ::Playwright::Error => err
          case err.message
          when /Element is not attached to the DOM/,
            /Execution context was destroyed, most likely because of a navigation/,
            /Cannot find context with specified id/,
            /Unable to adopt element handle from a different document/
            # ignore error for retry
            puts "[WARNING] #{err.message}"
            puts BC.clean(err.backtrace)
            puts
          else
            raise
          end
        end

        assert_page_alive(retry_count: retry_count - 1, &block)
      end

    end
  end
end
```
